### PR TITLE
Added option to change Editor app

### PR DESCRIPTION
**Pull request by ocerman**
This PR adds option to change F4 Editor in the preferences screen.

### CODE REVIEW

File: src/GenesisPanelView.cpp [hunk 1]
```diff
@@ -1811,17 +1811,34 @@ void PanelView::Edit(CustomListItem *item)
...
 
-			BString tempstring;
-			tempstring.SetTo("StyledEdit \"");
-			tempstring << file.String();
-			tempstring << "\" &";
-			system(tempstring.String());
+			app = SETTINGS->GetEditorApp();
+			if (app.IsEmpty())
+			{
+				BAlert *alert = new BAlert("Genesis","You have to choose editor app in the Preferences screen","Okay","Open Preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```
Comment (humdingerb) [inline src/GenesisPanelView.cpp]: Suggestion, considering standard Haiku naming: Choose->Select (see FileTypes prefs e.g.), Okay->OK
```suggestion
				BAlert *alert = new BAlert("Genesis","You have to select an editor application in the Genesis preferences","OK","Open preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```

File: src/GenesisPanelView.cpp [hunk 2]
```diff
@@ -1811,17 +1811,34 @@ void PanelView::Edit(CustomListItem *item)
...
+				return;
+			}
+
+			file.SetToFormat("%s/%s", m_Path.String(), item->m_FileName.String());
+			command.SetToFormat("%s \"%s\"", app.String(), file.String());
 
+			int32 ret = system(command.String());
+			if (ret != 0)
+			{
+				BAlert *alert = new BAlert("Genesis","Error while opening Editor application.\nYou can change editor app in the Preferences screen.","Okay","Open Preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```
Comment (humdingerb) [inline src/GenesisPanelView.cpp]:
```suggestion
				BAlert *alert = new BAlert("Genesis","Error while opening the editor application.\nYou can change the editor app in the Gensis preferences.","OK","Open preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```

File: src/GenesisPreferencesWindow.cpp [hunk 3]
```diff
@@ -35,6 +38,11 @@ GenesisPreferencesWindow::GenesisPreferencesWindow(BLooper* looper, BWindow *mai
 	m_SymlinkedPaths = new BCheckBox("followsymlinks", "Keep symlinked directory paths when navigating", new BMessage(PREFERENCES_CHANGED));
 	m_AskOnExit = new BCheckBox("askonexit", "Ask on exit", new BMessage(PREFERENCES_CHANGED));
 
+	m_EditorsMenu = new BMenu("choose ...",B_ITEMS_IN_COLUMN);
```
Comment (humdingerb) [inline src/GenesisPreferencesWindow.cpp]:
```suggestion
	m_EditorsMenu = new BMenu("Select ...",B_ITEMS_IN_COLUMN);
```

You can also replace "..." with B_UTF8_ELLIPSIS everywhere.

File: src/GenesisPreferencesWindow.cpp [hunk 4]
```diff
@@ -35,6 +38,11 @@ GenesisPreferencesWindow::GenesisPreferencesWindow(BLooper* looper, BWindow *mai
...
+	BMenuField *ChooseEditors = new BMenuField("",m_EditorsMenu);
+	ChooseEditors->SetDivider(0);
+
+	m_EditorApp = new BTextControl("editor", "F4 Editor application", "", NULL, B_WILL_DRAW|B_NAVIGABLE);
```
Comment (humdingerb) [inline src/GenesisPreferencesWindow.cpp]:
```suggestion
	m_EditorApp = new BTextControl("editor", "Editor application (F4)", "", NULL, B_WILL_DRAW|B_NAVIGABLE);
```

That way the "F4" isn't that prominent; the user may have set the option not to show function keys.

**Review by humdingerb — COMMENTED** [review]
It may be neat to just have the pop-up menu to select the editor app. The text field isn't really needed.
In that case it should be put at the top, before "Terminal window title". Looks better...

**Comment by ocerman** [comment]
Hello @humdingerb ,

Yes, propably the text field is not really needed. It would be usefull when Genesis would failed to detect some apps - but that seems to be working without any problems so far.

I will try to change it to pop-up menu in the future.

I have also accepted your other suggestions.

File: src/GenesisPanelView.cpp [hunk 5]
```diff
@@ -1811,17 +1811,34 @@ void PanelView::Edit(CustomListItem *item)
...
+				return;
+			}
+
+			file.SetToFormat("%s/%s", m_Path.String(), item->m_FileName.String());
+			command.SetToFormat("%s \"%s\"", app.String(), file.String());
 
+			int32 ret = system(command.String());
+			if (ret != 0)
+			{
+				BAlert *alert = new BAlert("Genesis","Error while opening the editor application.\nYou can change the editor app in the Gensis preferences.","OK","Open preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```
Comment (humdingerb) [inline src/GenesisPanelView.cpp]: Little typo :)
```suggestion
				BAlert *alert = new BAlert("Genesis","Error while opening the editor application.\nYou can change the editor app in the Genesis preferences.","OK","Open preferences",NULL,B_WIDTH_AS_USUAL,B_OFFSET_SPACING,B_STOP_ALERT);
```

**Review by humdingerb — CHANGES_REQUESTED** [review]

**Comment by humdingerb** [comment]
> Yes, propably the text field is not really needed. It would be usefull when Genesis would failed to detect some apps - but that seems to be working without any problems so far.

If needed, there could be an additional menu item "Other..." that opens a file dialog to pick an app manually, that is then added to the menu. When Genesis launches, it will then have to check if the app is still there and remove that menu item if it isn't.
But I suppose, it'll work well enough as it is, as long as the editor is correctly reporting its supported types etc.

**Comment by humdingerb** [comment]
Thanks!

**Pull request merged**